core/src/main/java/org/apache/iceberg/SupportsCommitValidation.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.function.BiConsumer;
+
+public interface SupportsCommitValidation {
+  CommitValidator NON_VALIDATING = (base, metadata) -> {};
+
+  @FunctionalInterface
+  interface CommitValidator extends BiConsumer<TableMetadata, TableMetadata> {
+    @Override
+    void accept(TableMetadata snapshot, TableMetadata snapshot2);
+  }
+
+  void commit(CommitValidator validator);
```

Comment:
Should we document that a  CommitFailedException should be thrown if the update fails validation? I'm not so sure as to specify that as a checked exception on the signature  since we don't do that elsewhere, but maybe something to indicate expectations to implementations.